language/move-vm/types/src/values/values_impl.rs

```diff
@@ -1811,6 +1832,91 @@ impl Vector {
     }
 }
 
+/***************************************************************************************
+ *
+ * Runtime Type checks
+ *
+ **************************************************************************************/
+
+impl Value {
+    pub fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        self.0.check_type(ty)
+    }
+}
+
+impl ValueImpl {
+    fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        match (self, ty) {
+            (ValueImpl::U8(_), Type::U8)
+            | (ValueImpl::U128(_), Type::U128)
+            | (ValueImpl::U64(_), Type::U64)
+            | (ValueImpl::Bool(_), Type::Bool)
+            | (ValueImpl::Address(_), Type::Address) => Ok(()),
+            (ValueImpl::IndexedRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.copy_value().read_ref()?.check_type(ty)
+            }
+            (ValueImpl::ContainerRef(r), Type::Reference(ty) | Type::MutableReference(ty)) => {
+                r.container().check_type(ty)
+            }
+            (ValueImpl::Container(c), ty) => c.check_type(ty),
+            (ValueImpl::U8(_), _)
+            | (ValueImpl::U128(_), _)
+            | (ValueImpl::U64(_), _)
+            | (ValueImpl::Bool(_), _)
+            | (ValueImpl::Address(_), _)
+            | (ValueImpl::IndexedRef(_), _)
+            | (ValueImpl::ContainerRef(_), _)
+            | (ValueImpl::Invalid, _) => Err(PartialVMError::new(
+                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+            )
+            .with_message(format!(
+                "unexpected type mismatch between {:?}, expected {:?}",
+                self, ty
+            ))),
+        }
+    }
+}
+
+impl Container {
+    fn check_type(&self, ty: &Type) -> PartialVMResult<()> {
+        match (self, ty) {
+            (Self::VecAddress(_), _)
+            | (Self::VecBool(_), _)
+            | (Self::VecU128(_), _)
+            | (Self::VecU64(_), _)
+            | (Self::VecU8(_), _) => Ok(()),
+            (Self::Vec(vs), Type::Vector(ty)) => {
+                for v in vs.borrow().iter() {
+                    v.check_type(ty)?;
+                }
```

Comment:
This is also very concerning. Here we're looping over the whole vector... Overall I feel like the recursive type check is equivalent to a deep copy, time complexity-wise.

Reply:
The reason why we are looping over vector is because we don't perform check for write_ref. If we don't do that, we can pass in a struct/vector that contains a resource that experienced a type hack. Lots of Move code right now are using reference to resource as authentication and if we missed that it would be a horrible error. Think about a passing to callee a`Vec<Capability>` where a value underneath is a `Foo`.

Reply:
With the new check on references, maybe we can omit this check.

Reply:
Definitely check_type + the signature token resolution are rather concerning regards performance.